drivers/gpio/gpio_nrfx.c

```diff
@@ -0,0 +1,467 @@
+/*
+ * Copyright (c) 2018, Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <gpio.h>
+#include <hal/nrf_gpio.h>
+#include <hal/nrf_gpiote.h>
+#include <string.h>
+
+#include "gpio_utils.h"
+
+struct gpio_nrfx_data {
+	sys_slist_t callbacks;
+
+	/* Mask holding information about which pins have been configured to
+	 * trigger interrupts using gpio_nrfx_config function.
+	 */
+	u32_t pin_int_en;
+
+	/* Mask holding information about which pins have enabled callbacks
+	 * using gpio_nrfx_enable_callback function.
+	 */
+	u32_t int_en;
+
+	u32_t active_level;
+	u32_t trig_edge;
+	u32_t double_edge;
+	u32_t inverted;
+};
+
+struct gpio_nrfx_cfg {
+	NRF_GPIO_Type *port;
+	u8_t port_num;
+};
+
+static inline struct gpio_nrfx_data *get_port_data(struct device *port)
+{
+	return port->driver_data;
+}
+
+static inline const struct gpio_nrfx_cfg *get_port_cfg(struct device *port)
+{
+	return port->config->config_info;
+}
+
+static int gpiote_channel_alloc(u32_t abs_pin, nrf_gpiote_polarity_t polarity)
+{
+	u8_t channel = GPIOTE_CH_NUM;
+
+	for (size_t i = 0; i < GPIOTE_CH_NUM; ++i) {
+		if ((NRF_GPIOTE->CONFIG[i] & GPIOTE_CONFIG_MODE_Msk)
+		    == GPIOTE_CONFIG_MODE_Disabled) {
+			channel = i;
```

Comment:
Why is it that complicated?
The loop should be `for (u8_t channel = 0; channel < GPIOTE_CH_NUM; ++channel)` and when an unused channel is found it should be configured and the function should return its index. Then after the loop only `return -ENODEV;` is sufficient.